1. Loading and Preprocessing
Loading the Dataset

In [51]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load the dataset
data = load_breast_cancer()
X = data.data  # Features
y = data.target  # Labels

Preprocessing Steps:
Handling Missing Values: The breast cancer dataset from sklearn is clean and does not contain any missing values. If there were missing values, we would use techniques like imputation (mean, median, or mode) or drop rows/columns with missing values.

Feature Scaling: The features in this dataset have different scales. Scaling is necessary for algorithms like SVM, k-NN, and Logistic Regression, which are sensitive to the magnitude of the features. We use StandardScaler to standardize the features to have a mean of 0 and a standard deviation of 1.

In [52]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Standardize the features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

Justification: Scaling ensures that all features contribute equally to the model's performance, preventing features with larger magnitudes from dominating those with smaller magnitudes.

2. Classification Algorithm Implementation:

    1.Logistic Regression
Description: Logistic Regression is a linear model used for binary classification. It estimates the probability of a sample belonging to a particular class using the logistic function.

Suitability: It is simple and interpretable, making it a good baseline model for binary classification tasks like this.

In [53]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Train the model
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

# Evaluate the model
y_pred = log_reg.predict(X_test)
print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred))

Logistic Regression Accuracy: 0.9824561403508771


2. Decision Tree Classifier
Description: A Decision Tree splits the data into subsets based on feature values, creating a tree-like structure to make predictions.

Suitability: It is easy to interpret and can handle non-linear relationships in the data.

In [54]:
from sklearn.tree import DecisionTreeClassifier

# Train the model
dtree = DecisionTreeClassifier(random_state=42)
dtree.fit(X_train, y_train)

# Evaluate the model
y_pred = dtree.predict(X_test)
print("Decision Tree Accuracy:", accuracy_score(y_test, y_pred))

Decision Tree Accuracy: 0.9415204678362573


3. Random Forest Classifier
Description: Random Forest is an ensemble method that builds multiple decision trees and combines their predictions to improve accuracy and reduce overfitting.

Suitability: It is robust to overfitting and performs well on high-dimensional datasets.

In [55]:
from sklearn.ensemble import RandomForestClassifier

# Train the model
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)

# Evaluate the model
y_pred = rf.predict(X_test)
print("Random Forest Accuracy:", accuracy_score(y_test, y_pred))

Random Forest Accuracy: 0.9707602339181286


4. Support Vector Machine (SVM)
Description: SVM finds the optimal hyperplane that separates the classes with the maximum margin.

Suitability: It is effective in high-dimensional spaces and works well for binary classification tasks.

In [56]:
from sklearn.svm import SVC

# Train the model
svm = SVC(kernel='linear', random_state=42)
svm.fit(X_train, y_train)

# Evaluate the model
y_pred = svm.predict(X_test)
print("SVM Accuracy:", accuracy_score(y_test, y_pred))

SVM Accuracy: 0.9766081871345029


5. k-Nearest Neighbors (k-NN)
Description: k-NN classifies a sample based on the majority class among its k-nearest neighbors in the feature space.

Suitability: It is simple and works well for small datasets with clear separation between classes.

In [57]:
from sklearn.neighbors import KNeighborsClassifier

# Train the model
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

# Evaluate the model
y_pred = knn.predict(X_test)
print("k-NN Accuracy:", accuracy_score(y_test, y_pred))

k-NN Accuracy: 0.9590643274853801


3. Model Comparison:

Algorithm	Accuracy:
Logistic Regression - 0.9825
Decision Tree - 0.9415
Random Forest - 0.9708
SVM - 0.9766
k-NN - 0.9591
Observations:
Best Performing Algorithm: Logistic Regression.

Worst Performing Algorithm: Decision Tree had the lowest accuracy.

Conclusion:
Logistic Regression is the best-performing model for this dataset, achieving the highest accuracy of 98.25%. It is a simple yet powerful algorithm for binary classification tasks like this.

SVM and Random Forest also performed very well, with accuracies of 97.66% and 97.08%, respectively. These models are suitable for datasets with complex relationships and high dimensionality.

k-NN performed decently with an accuracy of 95.91%, but it is slightly less effective than the top three algorithms.

Decision Tree performed the worst, with an accuracy of 94.15%, likely due to overfitting or suboptimal hyperparameters. However, it can be improved with proper tuning.